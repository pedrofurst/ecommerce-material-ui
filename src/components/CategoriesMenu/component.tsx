import { Button, List, ListItem, Popover, Typography } from '@material-ui/core';
import React, { useCallback, useState } from 'react';
import IconExpandMore from '@material-ui/icons/ExpandMore';
import IconExpandLess from '@material-ui/icons/ExpandLess';
import useStyles from './styles';

type CategoriesMenuPropsType = {
  categories: string[];
  selectedCategory: string;
  onSelectCategory: (category: string) => void;
};

function CategoriesMenu(props: CategoriesMenuPropsType) {
  const { categories, selectedCategory, onSelectCategory } = props;
  const classes = useStyles();
  const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null);

  const handlePopoverOpen = useCallback(
    (event: React.MouseEvent<HTMLElement, MouseEvent>) => {
      setAnchorEl(event.currentTarget);
    },
    []
  );

  const handlePopoverClose = useCallback(() => {
    setAnchorEl(null);
  }, []);

  const handleSelectCategory = useCallback(
    (category: string) => () => {
      onSelectCategory(category);
      handlePopoverClose();
    },
    [onSelectCategory, handlePopoverClose]
  );

  const open = Boolean(anchorEl);
  return (
    <>
      <Button
        className={classes.categoriesMenuContainer}
        color="inherit"
        onClick={handlePopoverOpen}
        id="categories-menu-button"
      >
        <Typography
          aria-owns={open ? 'mouse-over-popover' : undefined}
          aria-haspopup="true"
        >
          {selectedCategory || 'Categories'}
        </Typography>
        {open ? <IconExpandLess /> : <IconExpandMore />}
      </Button>

      <Popover
        open={open}
        anchorEl={anchorEl}
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'left',
        }}
        transformOrigin={{
          vertical: 'top',
          horizontal: 'left',
        }}
        onClose={handlePopoverClose}
        disableRestoreFocus
      >
        <List disablePadding>
          {categories.map((category: string) => (
            <ListItem
              key={category}
              button
              onClick={handleSelectCategory(category)}
              className="category-item"
            >
              <Typography className={classes.categoryText}>
                {category}
              </Typography>
            </ListItem>
          ))}
        </List>
      </Popover>
    </>
  );
}

export default CategoriesMenu;
