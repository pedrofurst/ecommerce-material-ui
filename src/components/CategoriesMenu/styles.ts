import { makeStyles } from '@material-ui/core/styles';

export default makeStyles({
  categoriesMenuContainer: {
    display: 'flex',
  },
  categoryText: {
    textTransform: 'capitalize',
  },
});
