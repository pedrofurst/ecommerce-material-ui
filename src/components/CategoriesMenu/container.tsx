import { useCallback } from 'react';
import { useHistory } from 'react-router-dom';
import useCategoryContext from '@features/providers/category/useCategoryContext';
import { ContainerIdType } from '@features/types/ContainerIdType';
import useProductContext from '@features/providers/product/useProductContext';
import CategoriesMenu from './component';

function CategoriesMenuContainer(props: ContainerIdType) {
  const {
    categories,
    selectedCategory,
    updateSelectedCategory,
  } = useCategoryContext();

  const { filterBy, clearFilter } = useProductContext();

  const history = useHistory();

  const handleSelectCategoryFilter = useCallback(
    (category: string) => {
      history.push('/');
      if (selectedCategory === category) {
        updateSelectedCategory('');
        clearFilter();
      } else {
        updateSelectedCategory(category);
        filterBy(category);
      }
    },
    [history, selectedCategory, updateSelectedCategory, clearFilter, filterBy]
  );

  return (
    <CategoriesMenu
      categories={categories}
      selectedCategory={selectedCategory}
      onSelectCategory={handleSelectCategoryFilter}
    />
  );
}

export default CategoriesMenuContainer;
