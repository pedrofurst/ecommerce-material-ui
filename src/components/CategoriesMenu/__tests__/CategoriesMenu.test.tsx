import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import CategoriesMenu from '../component';

describe('<CategoriesMenu />', () => {
  it('Should render categories menu', () => {
    const component = renderer.create(
      <CategoriesMenu
        categories={[]}
        onSelectCategory={jest.fn()}
        selectedCategory=""
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('Should render categories menu with categories', () => {
    const component = renderer.create(
      <CategoriesMenu
        categories={['Category1', 'Category2']}
        onSelectCategory={jest.fn()}
        selectedCategory=""
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
  it('Should render categories menu with selectedCategory', () => {
    const component = renderer.create(
      <CategoriesMenu
        categories={['Category1', 'Category2']}
        onSelectCategory={jest.fn()}
        selectedCategory="Category1"
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('Should call onSelectCategory when selects a category', () => {
    const onSelectCategory = jest.fn();
    const wrapper = shallow(
      <CategoriesMenu
        categories={['Category1']}
        onSelectCategory={onSelectCategory}
        selectedCategory=""
      />
    );
    wrapper
      .find('#categories-menu-button')
      .simulate('click', { currentTarget: {} });

    wrapper.find('.category-item').simulate('click');
    expect(onSelectCategory).toHaveBeenCalledWith('Category1');
  });
});
