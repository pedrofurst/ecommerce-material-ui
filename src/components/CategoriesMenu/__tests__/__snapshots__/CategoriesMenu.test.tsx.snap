// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoriesMenu /> Should render categories menu 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-categoriesMenuContainer-1 MuiButton-colorInherit"
  disabled={false}
  id="categories-menu-button"
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    <p
      aria-haspopup="true"
      className="MuiTypography-root MuiTypography-body1"
    >
      Categories
    </p>
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
      />
    </svg>
  </span>
</button>
`;

exports[`<CategoriesMenu /> Should render categories menu with categories 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-categoriesMenuContainer-1 MuiButton-colorInherit"
  disabled={false}
  id="categories-menu-button"
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    <p
      aria-haspopup="true"
      className="MuiTypography-root MuiTypography-body1"
    >
      Categories
    </p>
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
      />
    </svg>
  </span>
  <span
    className="MuiTouchRipple-root"
  />
</button>
`;

exports[`<CategoriesMenu /> Should render categories menu with selectedCategory 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-categoriesMenuContainer-1 MuiButton-colorInherit"
  disabled={false}
  id="categories-menu-button"
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    <p
      aria-haspopup="true"
      className="MuiTypography-root MuiTypography-body1"
    >
      Category1
    </p>
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
      />
    </svg>
  </span>
</button>
`;
